### PR TITLE
WIP/POC - Enable declarative way of panel options / editors definition (alternative version)

### PR DESCRIPTION
**Pull request by dprokop**
Alternative to #16731, based on that work as well

The idea behind this PR is to enable declarative way of panel editors definition.

## What we mean by declarative way of option editors definition?
This may be confusing, especially that React promises declarative way of building UI already. In this particular case we want make step further and simplify the way option editors are built.

The idea is to have an API that would enable editors construction without the need of knowing what @grafana/ui components to use when handling given option. Additionally, this API will enable possibility of creating DSLs around the options editors.

Let's assume for a minute that we are building Gauge panel options editor. The simplified `GaugeOptions` type looks as follows:

```ts
export interface GaugeOptions extends SingleStatBaseOptions {
  showThresholdLabels: boolean;
  showThresholdMarkers: boolean;
  thresholds: Threshold[];
}
```

For `thresholds` property we already have a unified editor, called `ThresholdEditor`. Building an edit for GaugeOptions should be a simple task then:

```ts
import { ThresholdEditor } from '@grafana/ui'
export class GaugePanelEditor extends PureComponent<PanelEditorProps<GaugeOptions>> {
  onThresholdsChange(thresholds) {
    const {options, onOptionsChange} = this.props;
    onOptionsChange({
      ...options,  
      thresholds,
  });

  render() {
    return (
        <div>
          {/*...*/}
          <ThresholdsEditor thresholds={this.props.options.thresholds} onChange={this.onThresholdsChange} />
           {/*...*/}
        </div>
    );
}
```

But it requires a lot of code, knowledge of the components APIs, `PanelEditorProps` API, etc.

We think there is a better way to do that by introducing and API for building Editors.

## The idea
The idea is fairly simple. Given an options type, let's refer to GaugeOptions again, we should be able to build the options UI without explicitly using editor components. To do so, we want to build the UI based on UI schema, that will be interpreted by set of components and rendered in the app.

```ts
// Example API, WIP

const uiSchema = new OptionsUIBuilder<GaugeOptions>();

uiSchema
  .addBooleanOption('showThresholdLabels')
  .addBooleanOption('showThresholdMarkers')
  .addThresholdsEditor('thresholds')
```

### Data validation
For data validation a yup object schema validator/builder could be used. Based on the data schema, the options editor builder could decide whether or not i.e. a given option is required, what are the validation requirements, etc.

### Structural layout of options object vs UI layout 
In multiple cases the structure of options object will not reflect the UI layout of the editor form. A good example of that are the latest changes in options model for single stats panel, where thresholds are part of [`FieldDisplayOptions`](https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/utils/fieldDisplay.ts#L21) interface. While having thresholds options defined in this interface might be thoughtful, then displaying the editor in the UI layout as a separate "box" also makes a lot of sense, especially of the UI nature of the thresholds editor.

To make it possible to build UI layouts around the options objects we also need to introduce some kind of model of "grouping" into the options UI schema. 

```ts
// Example API, WIP

const uiSchema = new OptionsUIBuilder<GaugeOptions>();

uiSchema
  .addOptionsGroup('Basic thresholds options')
    .addBooleanOption('showThresholdLabels')
    .addBooleanOption('showThresholdMarkers')
  .endGroup()
  .addGroup('Thresholds')
    .addThresholdsEditor('thresholds')
  .endGroup()
```

A very specific usage for grouping is the current panel options editor where options are organised in panels:
image

These panels are a specific implementation of options grouping, that introduce usage UI for the options:

```ts
// Example API, WIP

const uiSchema = new OptionsUIBuilder<GaugeOptions>();

uiSchema
  .addPanelOptionsGroup() // use panel options grid
    .addOptionsPanel('Basic thresholds options') // wrap options in options panel UI
      .addBooleanOption('showThresholdLabels')
      .addBooleanOption('showThresholdMarkers')
    .endGroup()
    .addOptionsPanel('Thresholds')
      .addThresholdsEditor('thresholds')
    .endGroup()
  .endGroup()
```

## Low level API
Low level API consists of two(three...) basic methods on the OptionsUIBuilder class:
1. `addGroup(config: GroupConfig)`
2. `addOptionEditor<TOption>(config: EditorConfig<TOption>)`
3. `addNestedOptionsGroup` (see Problems to be solved below)

GroupConfig and EditorConfig types are to be specified. The major part of the `EditorConfig` interface is the `component: React.ComponentType<OptionInputAPI<TOption>>` property that specifies what is the underlying primitive editor to be used for given option manipulation.

## Problems to be solved
- [ ] Editor components - the type requirement for option editors is `React.ComponentType<OptionInputAPI<TOption>>` type. *How to make it possible to configure additional props of the component that are not part of the OptionInputAPI?*  The very basic example is the `className` prop.
- [ ] Nested options objects
  Let's assume we have an options object of a type:
  ```ts
  interface MyOptions {
    a: string;
    b: { 
      c: number  
    }
  }```

  When defining the editor it would be ideal to be able to target the deep path properties in order to modify those, i.e.

  ```ts
  const uiSchema = new OptionsUIBuilder<MyOptions>();

  uiSchema
    .addOptionEditor('b.c', { useEditor: IntegerEditor);
  ```
  Unfortunately, this makes it impossible to infer the type of that particular option. Hence, the API user could use `BooleanEditor` to edit numeric option, which is inherently wrong, and should be prohibited on types level. To enable typing of nested options I propose following API:

```ts
  const uiSchema = new OptionsUIBuilder<MyOptions>();

  uiSchema
    .addNestedOptionsGroup('b') // editors take the  { c: number } type as a context in which API build editor
      .addOptionEditor('b.c', { useEditor: IntegerEditor)
   .endGroup()  // we are back in the top level context of options object 
```
  
This leads to "small" problem that I have encountered and is related to the type inference when ending a nested group. How to make TS infer the type of the higher level group? ( see https://github.com/grafana/grafana/pull/16773/files#diff-74e33ff61ee9655b16a4323902490904R17 for wrongly inferred type now)

### CODE REVIEW

**Comment by ryantxu** [comment]
quick look -- I think trying to use json-schema is worth the effort.  Having explicit constraints on the data model independent from the UI will be super helpful.

**Comment by dprokop** [comment]
> Having explicit constraints on the data model independent from the UI will be super helpful.

I agree now after playing with two separate schemas. And, according to the discussion we had, there is no need for schema <-> typings generation. With `OptionsDataSchema` type I have introduced it should be fairly type safe.

File: packages/grafana-ui/src/types/panelOptions.tsx [hunk 1]
```diff
@@ -0,0 +1,194 @@
+import { SelectOptionItem } from '../components/Select/Select';
+import { JSONSchema7 } from 'json-schema';
+
```
Comment (dprokop) [inline packages/grafana-ui/src/types/panelOptions.tsx]: Typings in this file are mess now, need general cleanup. Be gentle ;)

**Comment by torkelo** [comment]
Not really related to this just found this form framework that seems popular, might be something for our forms & validation 
https://jaredpalmer.com/formik/docs/overview

**Comment by dprokop** [comment]
@torkelo formik is a hell of a lib. Used it and really liked it.

**Comment by ryantxu** [comment]
If we like formik as a lib, we could consider yup for validation/schema rather than JSONschema.  I don't have any opinion, but am very supportive of trying to use external libraries that have already delt with these problems many many many times

**Comment by torkelo** [comment]
I do like the OptionsDataSchema that this has compared to the alternative. But I do miss the less verbose fluent API for defining the UI model. Specifying the UI model json schema manually looks very verbose and hard to read. But I assume that is what you mean by: 

> especially that the UI schema will be built using API (to come)

Looks like you put in a lot of work in these 2 PRs, very comprehensive foundation! Wonder how we should prioritize finalizing this work (to a mergable) state. Hate leaving so much work in a WIP state and moving on to other things (search etc).

File: public/app/plugins/panel/gauge/module.tsx [hunk 2]
```diff
@@ -1,10 +1,132 @@
...
+      type: 'number',
+      description: 'Hint for min value...',
+    },
+    showThresholdMarkers: {
+      type: 'boolean',
+    },
+    showThresholdLabels: {
+      type: 'boolean',
+    },
+    ...valueOptionsSchema.properties,
```
Comment (torkelo) [inline public/app/plugins/panel/gauge/module.tsx]: For the schema to be more useful we would need to extend it to include the higher level concepts like type: "unit", type: "valueOptions", intead of just the raw bools & strings.

File: packages/grafana-ui/src/types/panelOptions.tsx [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+export interface FieldsetUIModel
+  extends GroupLayoutUIModel<{
+    legend: string;
+  }> {} // legend property name inspired by https://www.w3schools.com/tags/tag_legend.asp
+
+/**
+ * Below are typeguards used by PanelOptionsBuilder components
+ */
+export function isOptionsUIModel(model: any): model is OptionsUIModel<any> {
+  return (model as OptionsUIModel<any>).model !== undefined;
```
Comment (torkelo) [inline packages/grafana-ui/src/types/panelOptions.tsx]: what is the purpose of this dynamic type check, why not have model property on OptionsUIModel?

**Comment by dprokop** [comment]
> Specifying the UI model json schema manually looks very verbose and hard to read
@torkelo this is the missing piece in this PR that I haven't managed to start on before vacation. The sketch for the API is in #16731

File: packages/grafana-ui/src/types/panelOptions.tsx (hunk 3 above)
Comment (dprokop) [reply packages/grafana-ui/src/types/panelOptions.tsx]: `OptionsUIModel` has model property on it. This is a type guard to decide whether or not the automatic options builder should be used. See https://github.com/grafana/grafana/pull/16773/files#diff-6e9276e76a52b87735ffbbb4306fa5a8R71

There's probably a better way to decide if the panel uses declarative options or not, but I wanted to use `editor` prop on panel now to iterate fast

**Comment by dprokop** [comment]
> Wonder how we should prioritize finalizing this work (to a mergable) state. Hate leaving so much work in a WIP state and moving on to other things (search etc).

@torkelo I agree. I see next steps here to be: 
- [ ] Prepare fluent API for primitive values and build i.e. UnitPicker out of it.
- [ ] Prepare components for primitive values (Number inputs, text, switch, etc...) to be aligned with `OptionInputAPI`. 
- [ ] Figure out way of handling select/multi value options
- [ ] Implement fluent API for reference for more complex options, i.e. UnitPicker (vide `.addUnitPicker`)

**Comment by torkelo** [comment]
Other tricky things to figure out.

* UI Layout vs structural layout  (we have some cases where the structure does not match UI layout). So might need deep "property" path syntax.
* Change to a more fluid layout. Think to have the UI model define columns is a bit too rigid, and something Grafana can calculate. Thinking that the groups can have "minWidth" properly (in 1-3) and then Grafana can layout the columns more intelligently. This should eliminate one nesting of the ui model I think. So it's just groups -> options.

File: public/app/plugins/panel/gauge/module.tsx (hunk 2 above)
Comment (dprokop) [reply public/app/plugins/panel/gauge/module.tsx]: I sort of agree. I think things like i.e. value options are in the end of `object` type with nested primitives. For data schema that will be helpful for validation I found Yup used by Formik for instance (https://github.com/jquense/yup) a nice way for data schema representation and validation. I haven't found similar API for JSON Schema though, that would help us build data schema without much effort.

Both approaches will enable us to to have those more abstract schemas, actually in this PR I have already naively tried that with: 

```ts
const valueOptionsSchema: ObjectOptionDataSchema<SingleStatBaseOptions> = {
  properties: {
    orientation: {},
    thresholds: {},
    valueMappings: {},
    valueOptions: {},
  },
};
```

that's ltr on composed in Gauge options schema:
```ts
  // ...
  properties: {
    minValue: {
      type: 'number',
      description: 'Hint for min value...',
    },
    maxValue: {
      type: 'number',
      description: 'Hint for min value...',
    },
    showThresholdMarkers: {
      type: 'boolean',
    },
    showThresholdLabels: {
      type: 'boolean',
    },
    ...valueOptionsSchema.properties,
  },
 // ...
```

In this case value options are not nested in Gauge options, which I think could be useful, especially when there are cases with lots of options that could be grouped (see react graph panel for reference https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/graph2/types.ts#L14)

There are of course JSON Schema validators that we could use for data validation (i.e. https://ajv.js.org/) but I haven't found anything close to Yup when it comes to how easy it is to both create and enable schema validation.

File: packages/grafana-ui/src/components/PanelOptions/OptionsUIBuilder.test.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+describe('OptionsUIBuilder', () => {
+    it('allows group as root', () => {
+      const builder = new OptionsGroupUIBuilder<GaugeOptions>();
+
+      const schema = builder
+          .addGroup({})
+            .addBooleanEditor('showThresholdLabels')
+            .addNestedOptionsGroup('fieldOptions', {})
```
Comment (dprokop) [inline packages/grafana-ui/src/components/PanelOptions/OptionsUIBuilder.test.ts]: @torkelo I have came up with this "fake grouping" for nested options. Type inferring based on property name works, yay, both in flat and nested scenario. But, see comments in the code below. Can't make the type after the group is ended (.groupEnd) be correct(it holds to nested option type down the rowad). Any ideas how the implementation of `endGroup` could be improved to handle that? Feels like I'm missing sth obvious here. cc @ryantxu

File: packages/grafana-ui/src/components/PanelOptions/OptionsUIBuilder.test.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+              .addThresholdsEditor('thresholds')
+            .endGroup() // How to return parent groupd ctx type here?
```
Comment (dprokop) [inline packages/grafana-ui/src/components/PanelOptions/OptionsUIBuilder.test.ts]:
```suggestion
            .endGroup() // How to return parent group ctx type here?
```

File: packages/grafana-ui/src/components/PanelOptions/OptionsUIBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+  getUIModel = () => {
+    return {
+      type: OptionsUIType.Editor as OptionsUIType.Editor, // TODO: how to fix this,
+      editor: this.model as any,
+    };
+  };
+}
+
+export class OptionsGroupUIBuilder<TOptions extends {}, TConfig extends {} = {}>
```
Comment (dprokop) [inline packages/grafana-ui/src/components/PanelOptions/OptionsUIBuilder.ts]: @torkelo @ryantxu that's the first version of the API for options ui schema building. No connection with yup schema yet.

File: packages/grafana-ui/src/types/panelOptions.ts [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+export enum OptionsUIType {
+  Layout = 'options:layout',
+  Group = 'options:group',
+  Editor = 'options:editor',
+}
+
+export interface OptionsUI<T> {
+  type: T;
+}
+export interface OptionsGrid extends OptionsUI<OptionsUIType.Layout> {
```
Comment (dprokop) [inline packages/grafana-ui/src/types/panelOptions.ts]: FYI I'm going to get rid of `OptionGrid`. So the UI model could accept groups and editors only. I'm thinking about Layout being a specific type of group that could be configured using OptionsGroup model, especially it accepts `component` that is used to render a group. 

If we want to implement i.e. PanelOptionsGrid layout using this API then we will add `.addPanelOptionsGrid` higher order method to `OptionsUIBuilder`, that will use `addGroup` underneath.

**Comment by torkelo** [comment]
Just found how visual studio code extensions define their settings: 
https://github.com/microsoft/vscode-tslint/blob/de22217e936fe508091592c3a8ae48707217f687/tslint/package.json#L50 

```
"configuration": {
      "type": "object",
      "title": "TSLint",
      "properties": {
        "tslint.enable": {
          "type": "boolean",
          "default": true,
          "description": "Control whether tslint is enabled for TypeScript files or not.",
          "scope": "resource"
        },
        "tslint.jsEnable": {
          "type": "boolean",
          "default": false,
          "description": "Control whether tslint is enabled for JavaScript files or not.",
          "scope": "resource"
        },
```

The thing that I found interesting is the "scope" property. Maybe that can be a better way to control the nesting, so we do not need to have everything on fieldOptions or field. Maybe we can have a scope property that can be "panel", "series" or "field" or something like that. 

thoughts @ryantxu? 

One thing that is a bit troubling with the recent fieldOptions & field changes (& open PRs that move thresholds and value mapping to field) is how nested the persisted options model become. And how this makes the UI & generic UI builder overly complicated. Having a flat option model would be a lot easier to work with.

**Comment by ryantxu** [comment]
+1 for something like the scope -- that seems like a good way to jump to the part of the stored values you are working with.

The `fieldOptions` structure is the most complicated one we are discussing and I think most of it can be custom react components.  

The declarative panel options will be most useful is defining the piles of simple inputs like:
image

**Pull request closed without merging**